packages/modules/vehicles/smarteq/api.py

```diff
@@ -60,60 +86,219 @@ def nested_key_exists(element: dict, *keys: str) -> bool:
+        self.password = conf.configuration.password
+        if self.password is None:
+            self.password = ""
+        self.pin = conf.configuration.pin
+        self.vin = conf.configuration.vin
+
+        if type(conf.configuration.useWebSocket) == bool:
+            self.useWebSocket = conf.configuration.useWebSocket
+        elif type(conf.configuration.useWebSocket) == str:
+            self.useWebSocket = json.loads(conf.configuration.useWebSocket)
+        else:
+            # if not properly defined set to False
+            self.useWebSocket = False
+
+        self.refreshToken = conf.configuration.refreshToken
+        self.refreshTokenMQTT = self.refreshToken
+
+        self.last_opmode = conf.configuration.opMode
+        if self.last_opmode is None:
+            self.last_opmode = 'unknown'
+        self.vehicle = vehicle
+        self.write_store_count = 0
+
+        self.su = socUtils()
+
         self.storeFile = str(RAMDISK_PATH) + '/soc_smarteq_store_vh_' + str(vehicle)
-        # LOGLEVEL = os.environ.get('LOGLEVEL', 'INFO').upper()
-        # logging.basicConfig(level=LOGLEVEL)
+        self.ws_storeFile = str(RAMDISK_PATH) + '/soc_smarteq_store_ws'
+        self.ws_messageFile = str(RAMDISK_PATH) + '/soc_smarteq_message_vh_' + str(vehicle)
 
-        # self.method keeps a high level trach of actions
+        if self.useWebSocket is True:
+            self.vhList = self.su.get_vin_ev_map('smarteq')
+            # self.log.debug("vehicle_vin_num_mapping=\n" + json.dumps(self.vhList, indent=4))
+            tn = self.ws_fname(self.vin)
+            if os.path.exists(tn):
+                self.load_ws_store(self.vin)
+            else:
+                self.init_ws_store(self.vin)
+                self.write_ws_store(self.vin)
+
+        # self.method keeps a high level trace of actions
         self.method = ''
         self.soc_ts = 'n/a'
         # self.store is read from ramdisk at start and saved at end.
-        # currently is contains:
-        # Tokens: refresh- and access-tokens of OAUTH
-        # refresh_timestamp: epoch of last refresh_tokens.
 
-        # self.session = requests.session()
         self.session = req.get_http_session()
 
+        self.init_store(False)
+        self.opmode = self.last_opmode
         self.load_store()
+        self.store['Tokens']['refresh_token'] = self.refreshToken
+
+        if self.password is not None and self.password != '':
+            self.opmode = 'pw-rest'
+        elif self.useWebSocket is False:
+            self.opmode = 'pin-rest'
+        else:
+            self.opmode = 'pin-ws'
+
+        if 'pin' in self.opmode and 'pin' in self.last_opmode:
+            self.log.debug('last and current opmode are pin-based, last:' + self.last_opmode +
+                           ', current:' + self.opmode)
+        else:
+            self.log.debug('switch opmode from ' + self.last_opmode + ' to ' + self.opmode)
+            self.store['opmode'] = self.opmode
```

Comment:
Die Klassen-Attribute, die es zusätzlich zu den Dict-Variablen gibt, machen das ganze sehr unübersichtlich. Ich will das an der Variable opmode verdeutlichen:
In Z 141 wird self.opmode auf self.last_opmode gesetzt. Dann in Z 146-15 wird self.opmode wieder neu gesetzt. Müsste nicht der last_opmode auf den opmode gesetzt werden? Um den alten Wert vor Beginn der Abfrage zu haben. Warum wird opmode nochmal neu initialisiert?

In Z. 157 wird self.opmode dann in self.store['opmode'] geschrieben.
In Z 293 wird self.opmode schließlich gepublished.
Für was wird die self.store['opmode'] benötigt?Wenn ich danach suche, wird darauf nie zugegriffen. Reicht nicht self.opmode, um den aktuellen Operation-Mode zu kennen? 

Zudem ist das ein großers Potenzial für Fehler. Haben self.opmode und self.store['opmode'] immer den gleichen Wert? Was passiert, wenn ich statt auf self.opmode auf self.store['opmode'] zugreife? Woher weiß ein anderer Entwickler, welches von beiden nun das richtige ist?